File: backend_node/src/domain/roomType/roomTypeName.ts
import { ValueObject } from "../../core/domain/ValueObject";
import { Guard } from "../../core/logic/Guard";
import { Result } from "../../core/logic/Result";

interface RoomTypeNameProps{
    value: string;
}

export class RoomTypeName extends ValueObject<RoomTypeNameProps>{
    get value(): string{
        return this.props.value;
    }

    private constructor(props: RoomTypeNameProps) {
        super(props);
    }

    public static create(roomTypeName: string): Result<RoomTypeName>{

        const guardResult = Guard.againstNullOrUndefined(roomTypeName,'roomTypeName')
        
        if (!guardResult.succeeded)
            return Result.fail<RoomTypeName>(guardResult.message);

        if (!roomTypeName.trim().length)  
            return Result.fail<RoomTypeName>('Room type name cannot be empty');

        return Result.ok<RoomTypeName>(new RoomTypeName({ value: roomTypeName }))
    }
}

File: dddnetcore/Domain/RoomTypes/RoomTypeName.cs
using System;
using DDDSample1.Domain.Shared;

namespace DDDSample1.Domain.RoomTypes{
    
    public class RoomTypeName{
        public String Name { get; private set; }

        public RoomTypeName(String name){

            if(string.IsNullOrEmpty(name))
                throw new BusinessRuleValidationException("The name of room type cannot be null or empty!");
            this.Name = name;
        }

        public override bool Equals(object obj){
            if (obj == null || GetType() != obj.GetType())
                return false;

            return Name.Equals(((RoomTypeName)obj).Name);
        }

        public override int GetHashCode()
        {
            return Name.GetHashCode();
        }
    }
}
